File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        n=len(nums)
        maxi=float("-inf")
        
        prefix=1
        for i in range(n):
            prefix*=nums[i]
            maxi=max(maxi,prefix)
            if prefix==0:
                prefix=1
        
        suffix=1
        for i in range(n-1,-1,-1):
            suffix*=nums[i]
            maxi=max(maxi,suffix)
            if suffix==0:
                suffix=1
        
        return maxi

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
       int maxi=nums[0];
       int mini = nums[0];
       int ans = nums[0];

       for(int i=1;i<nums.length;i++){
            if(nums[i]<0){
                int temp=maxi;
                maxi = mini;
                mini = temp;
            }

            maxi = Math.max(nums[i],maxi*nums[i]);
            mini = Math.min(nums[i],mini*nums[i]);

            ans = Math.max(maxi,ans);
       }

       return ans;
    }
}